User interface- Tictactoe project

ampaire opened the pull request:
Presents the simple user interface for the game.
@whiz25

ampaire committed:
Update changes

ampaire committed:
Update changes

ampaire committed:
Update changes

ampaire committed:
Update changes

ampaire committed:
Fix linting errors

davidmrtz-dev committed:
adding some user interfase methods

davidmrtz-dev committed:
adding methods to use rinterface

ampaire committed:
Update changes

davidmrtz-dev committed:
methods added user interface

davidmrtz-dev committed:
work on gitignore

davidmrtz-dev committed:
adding gitignore

ampaire committed:
Add file to the .gitignore file

davidmrtz-dev committed:
renaming licence file

stratospherique commented on bin/main.rb:
```diff
@@ -1 +1,72 @@
+
+  def take_place(inpt, play)
+    @board[inpt - 1] = play if play == 'O' || play == 'X'
+  end
```
- Board can be approached as a component (an object) which is independent from the DisplayInterface and has it's own behaviour (methods and attributes)

ampaire replied:
We have implemented the changes @stratospherique

stratospherique reviewed (changes requested):
### Great work team :clap: . However, some requirements needs to be implemented in order for us to approve your move to the next milestone. check below :arrow_down: 

* [x] `main.rb` uses `puts/print` and `get.chomp` commands.
  - It's the file where you are supposed to put all the code related to the game interactions with the user.
- **[REQUIRED BEHAVIOR]**
* [x] In this step not everything must work, the point is to make sure that you thought about everything listed below. You might do it only with inline comments.
* [x] Basic game flow is prepared:
  * [x] `[OPTIONAL]` game ask players for their names
  * [x] `[REQUIRED]` game tells which player turn it is
  * [x] `[REQUIRED]` game asks player to select from available moves
  * [x] `[REQUIRED]` game informs player if selected move is invalid
  * [x] `[REQUIRED]` game displays board after player move
  * [x] `[REQUIRED]` game informs player if selected move is a winning move
  * [x] `[REQUIRED]` game informs player if selected move is a draw move
  * [x] `[REQUIRED]` game repeats all actions for next player's move

##### Implementation details

* [x] A loop is needed in this game (recommended is `while` but anything else will do)
* [x] Inside loop, student should have all actions repeated for each move
* [x] A loop needs to be terminated at some point (after a winning move or draw to be more specific). It can be done with variable:

```ruby
game_on = true

# initial instructions for player

while game_on
  # loop for each move
  
  if winner # and/or draw (the exact condition in this milestone is not important)
    game_on = false
  end
end
```

### That's it :relieved: When finished, kindly resubmit another code review request :pray: 

_Happy coding_ :v: 
**Ahmed Mahfoudh**

davidmrtz-dev committed:
adding playing puts methods

davidmrtz-dev committed:
adding player turn method

davidmrtz-dev committed:
adding method requiered by TSE

davidmrtz-dev committed:
Adding some method requiered by TSE

davidmrtz-dev committed:
Texting output files

davidmrtz-dev commented:
Hi @stratospherique, thanks for the review... 👍 🔍 
We have already added the changes required, let me explain what did we do:
1. game ask players for their names => draw_welcome
2. game tells which player turn it is => draw_player_turn
3. game asks player to select from available moves => draw_available_moves
 4. game informs player if selected move is invalid => invalid_move
5. game displays board after player move =>draw_board
6. game informs player if selected move is a winning move =>winning_move
7. game informs player if selected move is a draw move =>draw_move
8. game repeats all actions for next player's move => We use a counter to show the methods but it will be controlled by a variable like you suggest:

game_on = true
#initial instructions for player
while game_on
  #loop for each move
  if winner # and/or draw (the exact condition in this milestone is not important)
    game_on = false
  end
end

Also thanks for the suggestions.
David Garcia. 💻 🤖

davidmrtz-dev committed:
adding methods to show the functionality

davidmrtz-dev committed:
added ds store to ignore

maelfosso commented on bin/main.rb:
```diff
@@ -1 +1,123 @@
+  game.draw_board
+  game.draw_line
+  counter += 1
+end
```
As indicated by @stratospherique , that loop has a problem. 
Please, implement what is suggest to you
```ruby
game_on = true

# initial instructions for player

while game_on
  # loop for each move
  
  if winner # and/or draw (the exact condition in this milestone is not important)
    game_on = false
  end
end
```

maelfosso commented on bin/main.rb:
```diff
@@ -1 +1,123 @@
+game.draw_available_moves(left_positions)
+game.invalid_move
+game.winning_move
+game.draw_move
```
Sorry, It's not the good place.

The **invalid move** and the **winning move** are detected into the **while loop**.

maelfosso commented on bin/main.rb:
```diff
@@ -1 +1,123 @@
+  # loop for each move
+  if game.get_current_player == player_one
+    game.draw_player_turn(player_one)
+    position = game.get_position.to_i
```
There are missing code here
```ruby
game.draw_player_turn(player_one)
position = game.get_position.to_i

# check of invalid move
# if the move is valid game.take_place otherwise loop until the move is valid
```

maelfosso commented on bin/main.rb:
```diff
@@ -1 +1,123 @@
+    position = game.get_position.to_i
+    game.take_place(position, 'O')
+  end
+  game.draw_board
```
Before **game.draw_board**, you need to check whether or not there is a win configuration in the board

maelfosso reviewed (changes requested):
# Changes required

Hello @ampaire @davitomix 

Please, you need to make the changes suggested by @stratospherique and read the comments to know more about the changes need to do.
Also, Kindly make your **bin/main.rb** file **executable**.

Once the changes done, please, submit a new code review request.
Thanks

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

davidmrtz-dev committed:
working on changes required

davidmrtz-dev committed:
adding changes on while loop according to the requeriments

davidmrtz-dev committed:
adding the changes to che wining and draw moves

davidmrtz-dev committed:
moved draw and win methods to ask before take the place

davidmrtz-dev committed:
adding sys generated file to gitignore

ampaire commented:
Hi @maelfosso, thanks for the review... 
We have already added the changes required
-We implemented the all the feedback :-
- Modified the while loop
- Moved the  **invalid move** and **winning move** to the while loop
- Put a check for the win configuration before the draw
- Added a check for the invalid move

Thank you for the review

stratospherique commented on bin/main.rb:
```diff
@@ -1 +1,150 @@
+    end
+  end
+  game.draw_line
+end
```
- Since you used a class for user interface, it's better to include the loop inside the userinterface class as a method and call it after initializing its object.
- The logic of the game loop should give the game the following features:
  - Ask the player for a move, if it's valid, put the symbol on the board, else it should ask again for a valid move. Use conditionals to achieve that.
  - For each valid move, check for the win or the draw for each player and display a message for the concerned player.
   - If the move is valid, it should do the same with the other player (i.e it goes on like that)

I noticed these flaws in the game loop after and changed to true each game condition methods :red_circle: separately.

ampaire replied:
- We put the while loop inside a method and called it after initializing the game object.

- However the logic that you are asking us to implement is a requirement of the game logic and we already implemented that in the game logic

stratospherique reviewed (changes requested):
### Great progress Team :clap: The game flow is almost done. However, I added some code comments below that you need to address in order for us to approve this milestone. :arrow_down: 

### That's it When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
_**Ahmed Mahfoudh**_

ampaire committed:
Add the whileloopinside the method

ampaire committed:
Implement changes

uche-inyama reviewed (approved):
## Hi team,  :wave:

I will go ahead to approve this milestone, I believe you have understood the flow of the game, which you showed in the comments, in main.rb, and is the primary purpose of the milestone. For reference purposes, use to the first comment from TSE - Ahmed.

However, for the next milestone, remember that, the methods for game should have it's own class, and therefore required from lib folder use this **require_relative ' ../lib/game_logic'** in your main.rb file. You should have an instance of game in your main.rb file, then add it within the while loop in your main.rb file
where necessary to achieve your flow.

#### ``` Have a great day ``` :slightly_smiling_face: 

``` Uchechukwu Henry Inyama```
